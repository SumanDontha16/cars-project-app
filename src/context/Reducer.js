/* eslint import/no-anonymous-default-export: [2, {"allowArrowFunction": true}] */
export default (state, action) => {
	switch (action.type) {
		case "CREATE_NEW_CAR":
			return {
				...state,
				car: [...state.car, action.payload],
			};

		case "UPDATE_CAR":
			const editingCar = action.payload;
			const updatedCar = state.car.map((car) => {
				if (car.id === editingCar.id) {
					return editingCar;
				}
				return car;
			});
			return {
				...state,
				car: updatedCar,
			};

		case "DELETE_CAR":
			return {
				...state,
				car: state.car.filter((car) => car.id !== action.payload),
			};

		default:
			return state;
	}
};
