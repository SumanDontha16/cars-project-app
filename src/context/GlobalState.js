import React, { createContext, useReducer } from "react";
import Reducer from "./Reducer";

const initialState = {
	car: [
		{
			id: 1,
			make: "Ford",
			model: "Puma",
			colour: "Black",
			year: "2009",
		},
	],
};

export const GlobalContext = createContext(initialState);
export const GlobalProvider = ({ children }) => {
	const [state, dispatch] = useReducer(Reducer, initialState);

	const createNewCar = (car) => {
		dispatch({
			type: "CREATE_NEW_CAR",
			payload: car,
		});
	};

	const deleteCar = (id) => {
		dispatch({
			type: "DELETE_CAR",
			payload: id,
		});
	};

	const updateCar = (car) => {
		dispatch({
			type: "UPDATE_CAR",
			payload: car,
		});
	};

	return (
		<GlobalContext.Provider
			value={{
				car: state.car,
				createNewCar,
				deleteCar,
				updateCar,
			}}
		>
			{children}
		</GlobalContext.Provider>
	);
};
