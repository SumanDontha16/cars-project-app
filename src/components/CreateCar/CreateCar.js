import React, { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import { Link } from "react-router-dom";
import { GlobalContext } from "../../context/GlobalState";

const CreateCar = () => {
	const [make, setMake] = useState("");
	const [model, setModel] = useState("");
	const [colour, setColour] = useState("");
	const [year, setYear] = useState("");

	const { createNewCar, car } = useContext(GlobalContext);

	const navigate = useNavigate();

	const onSubmit = (e) => {
		e.preventDefault();

		const addNewCar = {
			id: car.length + 1,
			make,
			model,
			colour,
			year,
		};
		createNewCar(addNewCar);
		navigate("/");
	};

	return (
		<>
			<div className="container updateCar">
				<h3 className="display-6 mb-5" data-testid="title">
					Create Car
				</h3>
				<form onSubmit={onSubmit}>
					<div className="mb-3">
						<label htmlFor="make" className="form-label">
							Make
						</label>
						<input
							data-testid="make"
							type="text"
							required={true}
							className="form-control"
							id="make"
							value={make}
							onChange={(e) => setMake(e.target.value)}
						/>
					</div>
					<div className="mb-3">
						<label htmlFor="model" className="form-label">
							Model
						</label>
						<input
							type="text"
							required={true}
							className="form-control"
							id="model"
							value={model}
							onChange={(e) => setModel(e.target.value)}
						/>
					</div>
					<div className="mb-3">
						<label htmlFor="colour" className="form-label">
							Colour
						</label>
						<input
							type="text"
							required={true}
							className="form-control"
							id="colour"
							value={colour}
							onChange={(e) => setColour(e.target.value)}
						/>
					</div>
					<div className="mb-3">
						<label htmlFor="year" className="form-label">
							Year
						</label>
						<input
							type="number"
							required={true}
							className="form-control"
							id="year"
							value={year}
							onChange={(e) => setYear(e.target.value)}
						/>
					</div>
					<button
						type="submit"
						data-testid="addcar"
						className="btn btn-primary me-2"
					>
						Add Car
					</button>
					<Link to="/">
						<button type="button" className="btn btn-dark">
							Cancel
						</button>
					</Link>
				</form>
			</div>
		</>
	);
};

export default CreateCar;
