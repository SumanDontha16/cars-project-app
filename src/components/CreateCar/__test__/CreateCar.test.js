import React from "react";
import { render, screen } from "@testing-library/react";
import user from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import CreateCar from "../CreateCar";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { GlobalProvider } from "../../../context/GlobalState";

describe("Create Car", () => {
	it("Test to match snapshot of component", () => {
		const createCarTree = renderer
			.create(
				<MemoryRouter>
					<CreateCar />
				</MemoryRouter>
			)
			.toJSON();
		expect(createCarTree).toMatchSnapshot();
	});

	it("Should renders with correct title", () => {
		render(
			<MemoryRouter>
				<CreateCar />
			</MemoryRouter>
		);
		const headerEl = screen.getByTestId("title");
		expect(headerEl.textContent).toBe("Create Car");
	});

	test("should return a default error message for fields that don't have a value", () => {
		render(
			<MemoryRouter>
				<CreateCar />
			</MemoryRouter>
		);
		const requiredField = screen.getByTestId("make");
		expect(requiredField.required).toBe(true);
	});

	test("should call the onSubmit when provided", async () => {
		render(
			<GlobalProvider>
				<MemoryRouter>
					<CreateCar />
				</MemoryRouter>
			</GlobalProvider>
		);

		const make = screen.getByLabelText(/make/i);
		const model = screen.getByLabelText(/model/i);
		const colour = screen.getByLabelText(/colour/i);
		const year = screen.getByLabelText(/year/i);

		user.type(make, "ford");
		user.type(model, "puma");
		user.type(colour, "black");
		user.type(year, "2008");

		const button = screen.getByRole("button", { name: "Add Car" });
		user.click(button);
	});
});
