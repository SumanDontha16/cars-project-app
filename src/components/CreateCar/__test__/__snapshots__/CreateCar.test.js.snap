// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Car Test to match snapshot of component 1`] = `
<div
  className="container updateCar"
>
  <h3
    className="display-6 mb-5"
    data-testid="title"
  >
    Create Car
  </h3>
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="make"
      >
        Make
      </label>
      <input
        className="form-control"
        data-testid="make"
        id="make"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="model"
      >
        Model
      </label>
      <input
        className="form-control"
        id="model"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="colour"
      >
        Colour
      </label>
      <input
        className="form-control"
        id="colour"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="year"
      >
        Year
      </label>
      <input
        className="form-control"
        id="year"
        onChange={[Function]}
        required={true}
        type="number"
        value=""
      />
    </div>
    <button
      className="btn btn-primary me-2"
      data-testid="addcar"
      id="addcar"
      type="submit"
    >
      Add Car
    </button>
    <a
      href="/"
      onClick={[Function]}
    >
      <button
        className="btn btn-dark"
        id="cancel"
        type="button"
      >
        Cancel
      </button>
    </a>
  </form>
</div>
`;
